proposals/type-classes.md

```diff
@@ -237,60 +202,155 @@ fun addInts(a: Int, b: Int, with Monoid<Int>): Int = add(a, b)
+
+1. In the same file as the type class definition (interface-side implementation)
+2. In the same file as the type being implemented (type-side implementation)
+
+All other instances are orphan instances and are not allowed. See [Appendix A](#Appendix-A) for a modification to this proposal that allows for orphan instances.
+
+Additionally a type class implementation must not conflict with any other already defined type class implementations; for the purposes of checking this we use the normal resolution rules.
+
+### Interface-Side Implementations
+
+This definition site is simple to implement and requires to rules except that the instances occurs in the same package. E.g. the following implementation is allowed
 ```kotlin
-fun <A> add(a: A, b: A, with Monoid<A>): A = a.combine(b)
+package foo.collections
+
+extension interface Monoid<A> {
+   ...
+}
 ```
-Call site:
+
 ```kotlin
-extension class AddingInts {
-  fun addInts(): Int = add(1, 2)
+package foo.collections
+
+extension object : Monoid<String> {
+   ...
 }
 ```
-The compiler may choose the following order for resolving the evidence that a `Monoid<Int>` exists in scope.
 
-1. Look in the most immediate scope for declarations of `with Monoid<Int>` in this case the function `addInts`
+### Type-Side Implementations
+
+This definition site poses additional complications when you consider multi-parameter typeclasses.
 
-This will compile because the responsibility of providing `Monoid<Int>` is passed into the callers of `addInts()`:
 ```kotlin
-extension object AddingInts {
-  fun addInts(with Monoid<Int>): Int = add(1, 2)
+package foo.collections
+
+extension interface Isomorphism<A, B> {
+   ...
 }
 ```
 
-2. Look in the immediately outher class/interface scope for declarations of `with Monoid<Int>` in this case the class `AddingInts`:
 ```kotlin
-extension class AddingInts(with Monoid<Int>) {
-  fun addInts(): Int = add(1, 2)
+package data.foo
+
+data class Foo(...)
+extension class<A> : Isomorphism<Foo, A> {
+   ...
 }
 ```
-This will compile because the responsibility of providing `Monoid<Int>` is passed unto the callers of `AddingInts()`
 
-3. Look in the import declarations for an explicitly imported instance that satisfies the constrain `Monoid<Int>`:
 ```kotlin
-import intext.IntMonoid
-extension object AddingInts {
-  fun addInts(): Int = add(1, 2)
+package data.bar
+
+data class Bar(...)
+extension class<A> : Isomorphism<A, Bar> {
+   ...
 }
 ```
-This will compile because the responsibility of providing `Monoid<Int>` is satisfied by `import intext.IntMonoid`
 
-4. Fail to compile if neither outer scopes nor explicit imports fail to provide evidence that there is a constrain satisfied by an instance in scope.
+The above instances are each defined alongside their respective type definitions and yet they clearly conflict with each other. We will also run into quandaries once we consider generic types. We can crib some prior art from Rust<sup>1</sup> to help us out here.
```

Comment:
Could we give just always precendence to local definitions flagged as internal and bail otherwise?
What happens if the user includes two libraries both providing conflicting instances on top of his locally defined one?

Reply:
There *cannot* be two libraries with conflicting instances if the compile obeys these rules. Downstream libraries cannot define instances that conflict with the type-class-providing library and libraries can only define instances for their local types.

This is twice the definition locations as classical interfaces but without losing any of the coherency of classical interfaces.